src/multi_period.py

```diff
@@ -0,0 +1,331 @@
+            verbose (bool): Flag to enable additional output.
+        """
+        super().__init__(
+            stocks=stocks,
+            budget=budget,
+            t_cost=t_cost,
+            bin_size=bin_size,
+            gamma=gamma,
+            file_path=file_path,
+            dates=dates,
+            model_type=model_type,
+            time_limit=time_limit,
+            alpha=alpha,
+            baseline=baseline,
+            sampler_args=sampler_args,
+            verbose=verbose,
+        )
+
+    def run(self, max_risk=0, min_return=0, num=0, init_holdings=None):
+        """Solve the rebalancing portfolio optimization problem.
+
+        Args:
+            max_risk (int): Maximum risk for the CQM risk bounding formulation.
+            min_return (int): Minimum return for the CQM return bounding formulation.
+        """
+        if not self.dates:
+            self.dates = DATES_DEFAULT
+
+        self.load_data()
+
+        num_months = len(self.df_all)
+        first_purchase = True
+        baseline_result = {}
+        self.baseline_values = [0]
+        self.update_values = [0]
+        months = []
+        initial_budget = 0
+
+        # Define dataframe to save output data
+        headers = ["Date", "Value"] + self.stocks + ["Variance", "Returns"]
+        self.opt_results_df = pd.DataFrame(columns=headers)
+        all_solutions = {}
+
+        self.price_df = pd.DataFrame(columns=self.stocks)
+
+        # Initialize the plot
+        plt.ylim(ymax=1.5 * self.budget, ymin=-1.5 * self.budget)
+        plt.xticks(
+            list(range(0, num_months, 2)),
+            self.df_baseline.index.strftime("%b")[::2],
+            rotation="vertical",
+        )
+        plt.locator_params(axis="x", nbins=num_months / 2)
+        plt.plot(
+            list(range(0, num_months)),
+            [0] * (num_months),
+            color="red",
+            label="Break-even",
+            linewidth=0.5,
+        )
+
+        for i in range(3, num_months):
+
+            baseline_result, months, all_solutions, init_holdings, initial_budget = self.run_update(
+                i,
+                first_purchase=first_purchase,
+                initial_budget=initial_budget,
+                baseline_result=baseline_result,
+                months=months,
+                all_solutions=all_solutions,
+                max_risk=max_risk,
+                min_return=min_return,
+                init_holdings=init_holdings,
+                cli_run=True,
+            )
+
+            first_purchase = False
+
+        print(self.opt_results_df)
+
+        plt.savefig("portfolio.png")
+
+        return all_solutions
+
+    def initiate_run_update(
+        self,
+        i: int,
+        baseline_result: dict,
+        months: list,
+        all_solutions: dict,
+        first_purchase: bool = True,
+        initial_budget: float = 0,
+        init_holdings: list = None,
+    ):
```

Comment:
(minor) Missing return type hint as well as `Optional` for `init_holdings`. Same for method below,